### PR TITLE
Jersey2 supports additional properties with composed schema

### PR DESCRIPTION
**Pull request by sebastien-rosset**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [x] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### CODE REVIEW

**Comment by sebastien-rosset** [comment]
@wing328

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java [hunk 1]
```diff
@@ -1646,10 +1646,24 @@ public void setAdditionalModelTypeAnnotations(final List<String> additionalModel
...
+        if (!supportsAdditionalPropertiesWithComposedSchema) {
+            // The additional (undeclared) propertiees are modeled in Java as a HashMap.
+            // 
+            // 1. supportsAdditionalPropertiesWithComposedSchema is set to false:
+            //    The generated model class extends from the HashMap. That does not work
+            //    with composed schemas that also use a discriminator because the model class
+            //    is supposed to extend from the generated parent model class.
+            // 2. supportsAdditionalPropertiesWithComposedSchema is set to true:
+            //    The HashMap is a field.
+            super.addAdditionPropertiesToCodeGenModel(codegenModel, schema);
```
Comment (sebastien-rosset) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java:1658]: @wing328 , this code avoids having the generated class extends from HashMap<String, T>. Instead, a `additionalProperties` field is added.

File: modules/openapi-generator/src/main/resources/Java/libraries/jersey2/additional_properties.mustache [hunk 2]
```diff
@@ -0,0 +1,31 @@
+{{#additionalPropertiesType}}
+  /**
+   * A container for additional, undeclared properties.
```
Comment (sebastien-rosset) [inline modules/openapi-generator/src/main/resources/Java/libraries/jersey2/additional_properties.mustache:3]: @wing328 , this is the code that has the `additionalProperties` field

File: modules/openapi-generator/src/main/resources/Java/libraries/jersey2/anyof_model.mustache [hunk 3]
```diff
@@ -33,9 +37,23 @@ public class {{classname}} extends AbstractOpenApiSchema{{#vendorExtensions.x-im
...
+                deserialized = tree.traverse(jp.getCodec()).readValueAs(cls);
+                {{classname}} ret = new {{classname}}();
+                ret.setActualInstance(deserialized);
+                return ret;
+            }
+            {{/discriminator}}
             {{#anyOf}}
             // deserialzie {{{.}}}
             try {
+                // TODO: custom deserialization when additionalProperties is set.
```
Comment (sebastien-rosset) [inline modules/openapi-generator/src/main/resources/Java/libraries/jersey2/anyof_model.mustache:56]: @wing328 , here we need to deserialize into the class, but we need to ignore the unknown properties and put them into the `additionalProperties` map.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java (hunk 1 above)
Comment (wing328) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java:1658]: @sebastien-rosset I may have a better way for this. Let's have a chat via Slack today.

File: modules/openapi-generator/src/main/resources/Java/libraries/jersey2/anyof_model.mustache (hunk 3 above)
Comment (wing328) [reply modules/openapi-generator/src/main/resources/Java/libraries/jersey2/anyof_model.mustache]: 👌 Let me try to come up with a way to handle this use case.

**Pull request merged**